Barco clickshare

Yarny0 opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Let NixOS users share their desktops with Barco's ClickShare dongles during presentations etc. See the package's description and Barco's homepage for more information on the technology.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS: x86_64-linux, x86_32-linux
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) ( *does not apply* )
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"` ( *does not apply* )
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after): *the package requires 5.9 MiB*
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Infinisil commented on nixos/modules/misc/ids.nix:
```diff
@@ -638,6 +638,7 @@
       rss2email = 312;
       cockroachdb = 313;
       zoneminder = 314;
+      clickshare = 315;
```
No need to use a static gid, just define the group like `users.groups.clickshare = {}` and it will automatically be assigned a gid

Yarny0 replied:
Cool, thanks, I didn't know that.
Branch got updated.

flokli commented on pkgs/applications/video/clickshare-csc1/default.nix:
```diff
@@ -0,0 +1,83 @@
+    throw "unsupported system: ${stdenv.hostPlatform.system}";
+  inherit usersGroup;
+  sourceRoot = ".";
+  postUnpack = builtins.readFile ./post-unpack.sh;
```
please inline these files.

flokli commented on pkgs/applications/video/clickshare-csc1/default.nix:
```diff
@@ -0,0 +1,83 @@
+    xorg.libXinerama
+    xorg.libXtst
+  ];
+  rpmArch =
```
please set this in a let expression inside `postUnpack`, and use ${rpmArch} in there.

flokli commented on nixos/modules/programs/clickshare.nix:
```diff
@@ -0,0 +1,27 @@
+      # The package brings a udev rule file with a group name.
+      # This module has to define that group.
+      # We check for a name mismatch here:
+      assert clicksharePkg.usersGroup=="clickshare";
```
Why do we check this here?
How can a user of the module redefine `usersGroup` by accident? :confused:

Yarny0 committed:
clickshare-csc1: init at 01.07.00.033

Barco ClickShare is a wireless presentation system
where a USB dongle transmits to a base station
that is connected with a beamer.

Note that this package also contains a udev rule
file to grant access to the dongle if connected.
Access is limited to a group "clickshare"
(the name may be altered by package options).

Yarny0 committed:
nixos/clickshare: init module

The clickshare-csc1 package brings a udev rule file
to grant access to the ClickShare dongle if connected.
This module provides an option to install that rule file.
Only users in the "clickshare" users group have access.

Yarny0 commented:
Thanks, @flokli, for the review and your suggestions. I updated my branch accordingly. Regarding the group name check, I'm just used to adding lots of assertions to catch any mismatches. I removed it.

I also rebased on current `nixos-unstable` channel and checked the 32bit version.

Yarny0 commented:
Updated the branch following a suggestion of #61312: Remove empty line at beginning and end of `stdenv.mkDerivation` argument. Also rebased on current `nixos-unstable`.

flokli commented on nixos/modules/programs/clickshare.nix:
```diff
@@ -0,0 +1,16 @@
+{
+
+  options.programs.clickshare-csc1.enable =
+    lib.options.mkEnableOption "Barco ClickShare CSC-1 driver/client";
```
Probably we should add a line that users need to be added to the `clickshare` group to be able to use it.

Yarny0 replied:
Done (I hope the interpunctation of `mkEnableOption` adds up correctly in the documentation).

flokli commented:
@Yarny0 , I can't really test this (as I don't have the hardware around), but I assume you did :-)

flokli commented:
Thanks!

Yarny0 commented:
For the record: Most of my tests happened on the release-19.03 branch and on a NixOS 19.03 machine. Once I also built it on top of the unstable branch and successfully used the binaries, also on a NixOS 19.03 machine, however. So the only thing I never tested is whether the udev rule recognizes the dongle on the unstable branch. But I'm optimistic about that :-)

Thanks for your help and all your efforts, @flokli !!
